rust/arrow/src/array/equal.rs

```diff
@@ -43,11 +43,15 @@ pub trait ArrayEqual {
 }
 
 impl<T: ArrowPrimitiveType> ArrayEqual for PrimitiveArray<T> {
-    default fn equals(&self, other: &dyn Array) -> bool {
+    fn equals(&self, other: &dyn Array) -> bool {
         if !base_equal(&self.data(), &other.data()) {
             return false;
         }
 
+        if T::DATA_TYPE == DataType::Boolean {
```

Comment:
Here we use the same principle: this will be a constant `if a == a` or `if a == b`.